src/vmm/src/version_map.rs

```diff
@@ -34,13 +42,14 @@ lazy_static! {
 
     /// Static instance used for creating a 1:1 mapping between Firecracker release version
     /// and snapshot data format version.
-    pub static ref FC_VERSION_TO_SNAP_VERSION: HashMap<String, u16> = {
+    pub static ref FC_VERSION_TO_SNAP_VERSION: HashMap<FcVersion, u16> = {
         let mut mapping = HashMap::new();
+        // aarch64 snapshotting support was introduced in v0.24.0
         #[cfg(not(target_arch = "aarch64"))]
-        mapping.insert(String::from("0.23.0"), 1);
+        mapping.insert(FC_VERSION_0_23_0, 1);
```

Comment:
What problem is replacing the string with FcVersion ? The extended version checks can be easily done against the data version. IMO this looks a bit over engineered or I am missing something.

Reply:
1. It looks like the natural thing is to perform the checks based on Fc version. In the end it is a Firecracker release that adds support for different features, and not a data version. Moreover, if you base the checks on the data version, you also have to store the reverse mapping. For example as it is done right now: `const FC_V0_23_SNAP_VERSION: u16 = 1;`. 
2. The `FcVersion` struct is very light and some of the logic included there is done anyway (for example checking if the fc_version string is valid). 

3. Consider the following situation:

```
    pub static ref FC_VERSION_TO_SNAP_VERSION: HashMap<String, u16> = {
        let mut mapping = HashMap::new();
        mapping.insert(String::from("0.23.0"), 1);
        mapping.insert(String::from("0.24.0"), 2);
        mapping.insert(String::from("0.25.0"), 3);
        mapping.insert(String::from("0.26.0"), 4);
        mapping.insert(String::from("0.27.0"), 5);

        mapping
    };
```

Now let's say that you want to add version `0.23.1`. You will probably do `mapping.insert(String::from("0.23.1"), 6);`.

And let's say that you want to perform some logic for Fc versions <= 0.25.0. You can do `data_version <= 3 && version != 6`, but these kind of checks can be hard to follow.

Reply:
The theoretical example makes sense and the yes, the solution is logical in that context. The complexity of managing state changes for patch releases makes it something we want to avoid entirely - in your example, Firecracker v0.24.0 would basically need to know how to load the v0.23.1 state which was mutated after v0.24.0 is released. My expectation and suggestion is to only do state changes as part of new release and never otherwise. This a

Reply:
It's true that managing state changes for patch releases is quite complex, but I wouldn't completely dismiss the possibility. I would postpone the decision until we face a situation where we need to take this into consideration.

In any case personally I think it would be better to perform snapshot creation checks on Fc version and not on the data version. Also the snapshotting API (`snapshot/create`) uses the Fc version and not the data version.

But if you have strong concerns about this, I can remove the `FcVersion` struct and try to redo the PR, keeping any simplifications that make sense without it.